helix-term/src/ui/editor.rs

```diff
@@ -891,14 +892,18 @@ impl EditorView {
                 for _ in 0..cxt.editor.count.map_or(1, NonZeroUsize::into) {
                     // first execute whatever put us into insert mode
                     self.last_insert.0.execute(cxt);
+                    let mut last_savepoint = None;
+                    let mut last_request_savepoint = None;
                     // then replay the inputs
                     for key in self.last_insert.1.clone() {
                         match key {
                             InsertEvent::Key(key) => self.insert_mode(cxt, key),
                             InsertEvent::CompletionApply(compl) => {
                                 let (view, doc) = current!(cxt.editor);
 
-                                doc.restore(view);
+                                if let Some(last_savepoint) = last_savepoint.as_deref() {
+                                    doc.restore(view, last_savepoint);
```

Comment:
Since this is just getting whatever the currently focused view is, that may not be the same as the view in the last save point, right? E.g. would this panic if we sent a completion request and then changed views?

Reply:
Your general idea is correct, that's why I added a check to drop requests if the view or document has changed in https://github.com/helix-editor/helix/pull/6173/commits/56bf37a5b55ac46bda51f85c8840a355bb113db3

Reply:
I think I'm not quite understanding how this works. It looks like those additional checks would prevent a new pop-up menu of completion items if we received a response after moving views. But this code looks like it's for handling repetition of a completion application, which in my understanding, does not trigger a new pop-up, just inserts the same text again. How does that extra check come into play here?

Reply:
The repetition code only replays events recorded while in insertmode and the appropriate event only gets generated if the completion menu is actually opened (prevent by said check). Specifically there are three events:

* Request completion: Generated immidietly when a completion is requested, generates a new savepoint when repeated
* trigger completion: Generated only if the completion menu is actually opened, active the savepoiny  described above (move it to a second variable)
* completion even: triggered when selecting something in the menu, restores the **active** savepoint

The important part is that there are always two savepoints during repeating: an active savepoint actually used for completions and the savepoint of the current request. 

So in the case you mentioned a completion request (and completion request event) would occur bit if the view/doc is switched then the request gets dropped after it finishes and no trigger completion event is ever generated. Therefore the savepoint belonging to the request is never activated and would be replaced by the next completion request event

Reply:
Got it, this makes a lot more sense now, thanks for the write-up!